samcli/commands/pipeline/init/interactive_init_flow.py

```diff
@@ -0,0 +1,244 @@
+            return expected_previous_clone_local_path
+        raise
+
+
+def _clone_custom_pipeline_template(repo_url: str) -> Path:
+    """
+    clone a given Git pipeline template's Git repo to the user machine in SAM shared directory.
+    Returns: the local directory path where the repo is cloned to.
+
+    Parameters:
+        repo_url: the URL of the Git repo to clone
+
+    Returns:
+        the local directory path where the repo is cloned.
+    """
+    return _clone_pipeline_templates(repo_url=repo_url, clone_name=CUSTOM_PIPELINE_TEMPLATE_REPO_LOCAL_NAME)
+
+
+def _clone_pipeline_templates(repo_url: str, clone_name: str) -> Path:
+    """
+    clone a given pipeline templates' Git repo to the user machine inside the SAM shared directory(default: ~/.aws-sam)
+    under the given clone name. For example, if clone_name is "custom-pipeline-template" then the location to clone
+    to is "~/.aws-sam/custom-pipeline-template/"
+
+    Parameters:
+        repo_url: the URL of the Git repo to clone
+        clone_name: The folder name to give to the created clone
+
+    Returns:
+        Path to the local clone
+    """
+    try:
+        repo: GitRepo = GitRepo(url=repo_url)
+        clone_path: Path = repo.clone(clone_dir=shared_path, clone_name=clone_name, replace_existing=True)
+        return clone_path
+    except (OSError, CloneRepoException) as ex:
+        raise PipelineTemplateCloneException(str(ex)) from ex
+
+
+def _read_app_pipeline_templates_manifest(pipeline_templates_dir: Path) -> PipelineTemplatesManifest:
+    """
+    parse and return the manifest yaml file located in the root directory of the SAM pipeline templates folder:
+
+    Parameters:
+        pipeline_templates_dir: local directory of SAM pipeline templates
+
+    Raises:
+        AppPipelineTemplateManifestException if the manifest is not found, ill-formatted or missing required keys
+
+    Returns:
+        The manifest of the pipeline templates
+    """
+    manifest_path: Path = Path(os.path.normpath(pipeline_templates_dir.joinpath("manifest.yaml")))
+    return PipelineTemplatesManifest(manifest_path=manifest_path)
+
+
+def _select_pipeline_template(pipeline_templates_manifest: PipelineTemplatesManifest) -> PipelineTemplateManifest:
+    """
+    Prompts the user a list of the available CI/CD providers along with associated app pipeline templates to choose
+    one of them
+
+    Parameters:
+        pipeline_templates_manifest: A manifest file lists the available providers and the associated pipeline templates
+
+    Returns:
+         The manifest (A section in the pipeline_templates_manifest) of the chosen pipeline template;
+    """
+    provider = _prompt_for_cicd_provider(pipeline_templates_manifest.providers)
+    provider_pipeline_templates: List[PipelineTemplateManifest] = list(
+        filter(lambda t: t.provider == provider, pipeline_templates_manifest.templates)
```

Comment:
It is better to use this form rather than `filter` and lambda function: `[something for something in some_list if something.xyz == abc]`

See this post: https://stackoverflow.com/questions/1247486/list-comprehension-vs-map/6407222#6407222